File: apps/backend/src/main/kotlin/org/eu/nl/craftsmansledger/technologyTrees/TechnologyTreesRepository.kt
package org.eu.nl.craftsmansledger.technologyTrees

import io.viascom.nanoid.NanoId
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.deleteWhere
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.statements.InsertStatement
import org.jetbrains.exposed.sql.transactions.transaction
import org.jetbrains.exposed.sql.update

fun ResultRow.toTechnologyTree() = TechnologyTree(
    this[TechnologyTreeTable.id],
    this[TechnologyTreeTable.name],
    this[TechnologyTreeTable.maxPoints],
)

fun InsertStatement<Number>.toTechnologyTree() = TechnologyTree(
    this[TechnologyTreeTable.id],
    this[TechnologyTreeTable.name],
    this[TechnologyTreeTable.maxPoints],
)

class TechnologyTreesRepository {
    fun findAll() = transaction {
        TechnologyTreeTable.selectAll().map { it.toTechnologyTree() }.toList().sortedBy { it.name }
    }

    fun findOneById(technologyTreeId: String) = transaction {
        TechnologyTreeTable
            .selectAll()
            .where { TechnologyTreeTable.id eq technologyTreeId }
            .map { it.toTechnologyTree() }
            .singleOrNull()
    }

    fun findOneByName(name: String) = transaction {
        TechnologyTreeTable
            .selectAll()
            .where { TechnologyTreeTable.name eq name }
            .map { it.toTechnologyTree() }
            .singleOrNull()
    }

    fun create(data: CreateTechnologyTreeData) = transaction {
        TechnologyTreeTable
            .insert {
                it[id] = NanoId.generate()
                it[name] = data.name
                it[maxPoints] = data.maxPoints
            }
            .toTechnologyTree()
    }

    fun update(data: TechnologyTree): TechnologyTree {
        transaction {
            TechnologyTreeTable.update({ TechnologyTreeTable.id eq data.id }) {
                it[name] = data.name
                it[maxPoints] = data.maxPoints
            }
        }
        return this.findOneById(data.id)!!
    }

    fun remove(technologyTreeId: String) {
        transaction { TechnologyTreeTable.deleteWhere { TechnologyTreeTable.id eq technologyTreeId } }
    }
}

val technologyTreesRepository = TechnologyTreesRepository()

File: apps/api/src/app/technology-trees/technology-trees.repository.ts
import { CreateTechnologyTreeData, serialize, serializeAll, TechnologyTree } from '@craftsmans-ledger/shared';
import { Injectable } from '@nestjs/common';
import { DatabaseService } from '../core';

@Injectable()
export class TechnologyTreesRepository {
    constructor(private readonly databaseService: DatabaseService) {}

    public async findAll() {
        const results = await this.databaseService.prismaClient.techTree.findMany();
        return serializeAll(TechnologyTree, results);
    }

    public async findOneById(technologyTreeId: string) {
        const result = await this.databaseService.prismaClient.techTree.findUnique({
            where: { id: technologyTreeId },
        });
        return serialize(TechnologyTree, result);
    }

    public async findOneByName(name: string) {
        const result = await this.databaseService.prismaClient.techTree.findUnique({
            where: { name: name },
        });
        return serialize(TechnologyTree, result);
    }

    public async create(data: CreateTechnologyTreeData) {
        const created = await this.databaseService.prismaClient.techTree.create({
            data: {
                name: data.name,
                maxPoints: data.maxPoints,
            },
        });
        return serialize(TechnologyTree, created);
    }

    public async update(data: TechnologyTree) {
        const updated = await this.databaseService.prismaClient.techTree.update({
            where: { id: data.id },
            data: {
                name: data.name,
                maxPoints: data.maxPoints,
            },
        });
        return serialize(TechnologyTree, updated);
    }

    public async remove(technologyTreeId: string) {
        await this.databaseService.prismaClient.techTree.delete({ where: { id: technologyTreeId } });
    }
}
